### ISSUE TITLE
SYSTEM:  Command browse_website returned: Error: Message: Service /home/appuser/.wdm/drivers/chromedriver/linux64/112.0.5615.49/chromedriver unexpectedly exited. Status code was: 127

### ISSUE THREAD

**Issue by psvensson** [Significant-Gravitas/Auto-GPT#1821]
### Duplicates

- [X] I have searched the existing issues

### Steps to reproduce 🕹

1. Build a dockerized version of aouto-gpt
2. Give it any instruction that leads to it needing to browse_website

### Current behavior 😯

SYSTEM:  Command browse_website returned: Error: Message: Service /home/appuser/.wdm/drivers/chromedriver/linux64/112.0.5615.49/chromedriver unexpectedly exited. Status code was: 127

### Expected behavior 🤔

Should be bale to browse website properly

### Your prompt 📝

```yaml
# It really could be any prompt
```

**Comment by lab34** [comment]
Hi, just my two cents:
I have the same problem, but with exit code 255. What is strange, is that if we go inside the running container, in /home/appuser, there is no .wdm directory , and if we try to find / -name chrome* in the container, we don't find chromedriver. I've tried to install chromium in the dockerfile, but no luck. I'm still searching...

**Comment by psvensson** [comment]
There was a discussion in the Discord #dev-autogpt channel and one suggestion was that it might be permission issues. The installation process might download the right thing but is not able to save it, or perhaps everything is installed fine, but the docker user is not allowed to either run it or access the files (or both)

**Comment by psvensson** [comment]
I have now built a new docker image of master with the USER command commented out, so the container user is root, and I get exactly the same error.

this means it is most probably not a security issue but something somewhere which is not installed or configured properly

**Comment by doem97** [comment]
Same here, running on Docker on MacOS.
```
Command browse_website returned: Error: Message: Service /home/appuser/.wdm/drivers/chromedriver/linux64/112.0.5615.49/chromedriver unexpectedly exited. Status code was: 255
```
I found that this error got no history under Issues. One possible solution may be to check recent git commits. (I guess should be limited within half a day)

**Comment by doem97** [comment]
WTH this repo got too many updates every single hour. Should be related to browse_website file?

**Comment by psvensson** [comment]
Heh yes maybe. OK, I have got some progress. There was some libraries missing. It might be related to a recent name change of the chromium-driver. Hope to solve this soon..

**Comment by psvensson** [comment]
By adding this to the Dockerfile, I get further;

RUN apt-get install -y libglib2.0 libnss3 libgconf-2-4 libfontconfig1 chromium-driver

The error is now;

WDM] - Downloading: 100%|██████████████████████....█████████████████████████████| 6.75M/6.75M [00:00<00:00, 15.9MB/s]

SYSTEM:  Command browse_website returned: Error: Message: unknown error: Chrome failed to start: exited abnormally. (unknown error: DevToolsActivePort file doesn't exist) (The process started from chrome location /usr/bin/chromium is no longer running, so ChromeDriver is assuming that Chrome has crashed.) Stacktrace: #0 0x5642ba4d1fe3 <unknown> #1.....

Note that the download succeeds - it downloads 6.75MB data. After that, the error occur.
I'm adding progress here in case another intrepid explorer does some research and can help further in case I get stuck.

**Comment by psvensson** [comment]
I think I found the problem. BY adding some options before starting the chrome webdriver, the problem goies away;

if CFG.selenium_web_browser == "firefox":
        driver = webdriver.Firefox(
            executable_path=GeckoDriverManager().install(), options=options
        )
    elif CFG.selenium_web_browser == "safari":
        # Requires a bit more setup on the users end
        # See https://developer.apple.com/documentation/webkit/testing_with_webdriver_in_safari
        driver = webdriver.Safari(options=options)
    else:
        options.add_argument('--no-sandbox')
        options.add_argument('--window-size=1420,1080')
        options.add_argument('--headless')
        options.add_argument('--disable-gpu')
        driver = webdriver.Chrome(
            executable_path=ChromeDriverManager().install(), options=options
        )
    driver.get(url)

That's the lines with 'option.add'

This is the first Python code I've ever written/pasted from Google btw :)
Will do a PR now..

**Comment by psvensson** [comment]
Solved by this PR https://github.com/Significant-Gravitas/Auto-GPT/pull/1857

**Comment by hemangjoshi37a** [comment]
Solution: 

The issue occurred due to missing libraries in the Dockerfile. The solution is to add the required libraries to the Dockerfile. 

```
RUN apt-get install -y libglib2.0 libnss3 libgconf-2-4 libfontconfig1 chromium-driver
```

Additionally, adding options to the Chrome webdriver also helped to resolve the issue. 

```
options.add_argument('--no-sandbox')
options.add_argument('--window-size=1420,1080')
options.add_argument('--headless')
options.add_argument('--disable-gpu')
```

These changes have been made in the PR #1857 and it has been merged. Please check if the issue is resolved by updating your code with the changes in the PR.

**Comment by lab34** [comment]
Thank you, 
no luck for me, I have the same error then before but know, but it's related with the Apple Silicon architecture now, 
`SYSTEM:  Command browse_website returned: Error: Message: Service /home/appuser/.wdm/drivers/chromedriver/linux64/112.0.5615.49/chromedriver unexpectedly exited. Status code was: 255
`
So, I connect to the terminal inside the container, and now I see the .wdm directory. If I'm going inside and launch the chromedriver, I have this error: 
```
$ ./chromedriver
qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
```
it send me to this: https://github.com/browserless/chrome/issues/2004
It's true that my Macbook is an M1, so it's another specific issue.
So, I've ticked "Use Rosetta..." in docker desktop, and now, when I launch chromedriver, I have 
`rosetta error: failed to open elf at /lib64/ld-linux-x86-64.so.2
Trace/breakpoint trap`
So I did
`export DOCKER_DEFAULT_PLATFORM=linux/amd64 `
And now I have the status code 127 back and
`./chromedriver: error while loading shared libraries: libglib-2.0.so.0: cannot open shared object file: No such file or directory
`
aaarg... :-( endless debugging...

**Comment by montanaflynn** [comment]
@p-i-  I also have same issue, also using docker on an apple M2 macbook...

```
SYSTEM:  Command browse_website returned: Error: Message: Service /home/appuser/.wdm/drivers/chromedriver/linux64/112.0.5615.49/chromedriver unexpectedly exited. Status code was: 255
```

**Comment by WangPeterXF** [comment]
> I think I found the problem. BY adding some options before starting the chrome webdriver, the problem goies away;
> 
> if CFG.selenium_web_browser == "firefox": driver = webdriver.Firefox( executable_path=GeckoDriverManager().install(), options=options ) elif CFG.selenium_web_browser == "safari": # Requires a bit more setup on the users end # See https://developer.apple.com/documentation/webkit/testing_with_webdriver_in_safari driver = webdriver.Safari(options=options) else: options.add_argument('--no-sandbox') options.add_argument('--window-size=1420,1080') options.add_argument('--headless') options.add_argument('--disable-gpu') driver = webdriver.Chrome( executable_path=ChromeDriverManager().install(), options=options ) driver.get(url)
> 
> That's the lines with 'option.add'
> 
> This is the first Python code I've ever written/pasted from Google btw :) Will do a PR now..

may i know which .py file you made this change? also new to coding

**Comment by templeman15** [comment]
This also does not work for me with the latest code on `stable`. 

@WangPeterXF the file he's referring to is `commands/web_selenium.py`

But the solution did not work for me.

**Comment by martoast** [comment]
I am having the same error on Macbook M1:

Service /home/appuser/.wdm/drivers/chromedriver/linux64/112.0.5615.49/chromedriver unexpectedly exited. Status code was: 255

The issue appeard to be that Chrome Driver is not correctly installed or configured.

**Comment by interwubs** [comment]
I too am on an M1.

I set a `--platform=linux/amd64` in the Dockerfile `FROM --platform=linux/amd64 python:3.10-slim` just to see if it would help.

Incredibly slow build, running wasn't too terrible.

```
SYSTEM:  Command browse_website returned: Error: Message: unknown error: Chrome failed to start: exited abnormally. (chrome not reachable) (The process started from chrome location /usr/bin/chromium is no longer running, so ChromeDriver is assuming that Chrome has crashed.) Stacktrace: #0 0x0040006e6fe3 <unknown> #1 0x004000425d36 <unknown> #2 0x00400044eb20 <unknown> #3 0x00400044aa9b <unknown> #4 0x00400048caf7 <unknown> #5 0x00400048c11f <unknown> #6 0x004000483693 <unknown> #7 0x00400045603a <unknown> #8 0x00400045717e <unknown> #9 0x0040006a8dbd <unknown> #10 0x0040006acc6c <unknown> #11 0x0040006b64b0 <unknown> #12 0x0040006add63 <unknown> #13 0x004000680c35 <unknown> #14 0x0040006d1138 <unknown> #15 0x0040006d12c7 <unknown> #16 0x0040006df093 <unknown> #17 0x004002585ea7 start_thread
```

Nope

**Comment by MnrGreg** [comment]
I've had good milage using Seleniarm (Selenium for ARM64) on M1
https://github.com/seleniumhq-community/docker-seleniarm


```Dockerfile
FROM seleniarm/standalone-chromium:latest

# Install git
USER root
RUN apt-get -y update
RUN apt-get -y install git pip

# Install Xvfb and other dependencies for headless browser testing
RUN apt-get update \
    && apt-get install -y wget gnupg2 libgtk-3-0 libdbus-glib-1-2 dbus-x11 xvfb ca-certificates

# Set environment variables
ENV PIP_NO_CACHE_DIR=yes \
    PYTHONUNBUFFERED=1 \
    PYTHONDONTWRITEBYTECODE=1

# Create a non-root user and set permissions
WORKDIR /home/seluser
RUN chown seluser:seluser /home/seluser
USER seluser

# Copy the requirements.txt file and install the requirements
COPY --chown=seluser:seluser requirements.txt .
RUN sed -i '/Items below this point will not be included in the Docker Image/,$d' requirements.txt && \
	pip install --no-cache-dir --user --break-system-packages -r requirements.txt

# Copy the application files
COPY --chown=seluser:seluser autogpt/ ./autogpt

# Set the entrypoint
ENTRYPOINT ["python3", "-m", "autogpt"]
```

**Comment by barrettj44** [comment]
In the file at location autogpt/commands/web_selenium.py, change line 90 to

driver = webdriver.Chrome( executable_path="/usr/bin/chromedriver",
options=options )

and re-build the Docker container. This resolved the issue for me

**Comment by hemangjoshi37a** [comment]
TLDR; [SOLUTION]

1. Add required libraries to the Dockerfile:
   ```
   RUN apt-get install -y libglib2.0 libnss3 libgconf-2-4 libfontconfig1 chromium-driver
   ```

2. Add options to the Chrome webdriver in the `web_selenium.py` file:
   ```
   options.add_argument('--no-sandbox')
   options.add_argument('--window-size=1420,1080')
   options.add_argument('--headless')
   options.add_argument('--disable-gpu')
   ```

except apple M1 chip.

**Comment by hemangjoshi37a** [comment]
@joserodolfofreitas As of now currently we dont have any issue tracking the behavior on Apple Silicon. But it is a nice idea to have

**Comment by yashb042** [comment]
@hemangjoshi37a  The solution you posted isn't working.

Machine details : Mac M1.

**Comment by hemangjoshi37a** [comment]
> except apple M1 chip

I already mentioned in my comment this. `except apple M1 chip`  @yashb042

**Comment by yashb042** [comment]
> > except apple M1 chip
> 
> I already mentioned in my comment this. `except apple M1 chip` @yashb042

My bad, didn't see the last line

**Comment by popbones** [comment]
I'm running the latest stable branch with Docker on an Apple Silicon Mac.

I'm able to load simple web pages with the following changes:

1. Added all the options mentioned above to `web_selenium.py` at L71 :
```
options.add_argument('--no-sandbox')
options.add_argument('--window-size=1420,1080')
options.add_argument('--headless')
options.add_argument('--disable-gpu')
```

2. Make sure firefox is selected as the browser in .env file.

3. This one probably is irrelevant, but I changed it at the beginning and don't have time to test w/o it:
I removed the arch specifier in the deb command when installing Chrome at L14 in Dockerfile. (Note that this doesn't seem to help with Chrome on M1 Macs either).

I only tested with some simple static web pages. It seems to be loading fine. Not sure if firefox would crash if a website uses some fancy web features that requires GPU acceleration.

**Comment by martinwunderlich** [comment]
Just for the record: Same isssue here (MacBook Pro 2021 with M1, Chrome version Version 112.0.5615.137). Error msg: 
`Command browse_website returned: Error: Message: Service /home/appuser/.wdm/drivers/chromedriver/linux64/112.0.5615.49/chromedriver unexpectedly exited. Status code was: 255`

**Comment by mawais01** [comment]
I am getting the same error on Mac M1 Pro 2021
`[ERROR] WebDriverException: Message: Service /root/.wdm/drivers/chromedriver/linux64/114.0.5735.90/chromedriver unexpectedly exited. Status code was: 255`

**Comment by mtcaddy** [comment]
Same here on Mac - Apple M2 Pro running in docker.
`SYSTEM:  Command browse_website returned: Error: unknown error: Chrome failed to start: exited abnormally.`
